standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/retryhandling/DataSourceWrapper.java

```diff
@@ -0,0 +1,287 @@
+   */
+  public Connection getConnection() {
+    return DataSourceUtils.getConnection(dataSources.get(RetryContext.getRetryContext().datasource));
+  }
+
+  /**
+   * @return Returns a {@link DatabaseProduct} instance representing the type of the underlying HMS dabatabe.
+   */
+  public DatabaseProduct getDatabaseProduct() {
+    return databaseProduct;
+  }
+
+  /**
+   * <p>Establishes a {@link RetryContext} on the current {@link Thread}: Creates a new transaction with the given 
+   * {@link TransactionDefinition} using an internal {@link PlatformTransactionManager} associated with the reffered 
+   * {@link DataSource}. Returns a {@link TransactionStatus} representing the created transaction. The returned 
+   * {@link TransactionStatus} instance can be used for manual transaction handling.</p> 
+   * <br>
+   * <p>!!!! Please note that if there already is a {@link RetryContext} established, this call will create a nested</p>
+   * {@link RetryContext} (and transaction).</b>  
+   * @param definition The {@link TransactionDefinition} to use for creating a new transaction.
+   * @param dataSource The identifier of the {@link DataSource} to use
+   * @return Returns a {@link TransactionStatus} object representing the created transaction.
+   * @throws TransactionException Forwarded from {@link PlatformTransactionManager#getTransaction(TransactionDefinition)}
+   */
+  public TransactionStatus getTransactionWithinRetryContext(TransactionDefinition definition, String dataSource) throws TransactionException {
+    TransactionStatus status = transactionManagers.get(dataSource).getTransaction(definition);
+    RetryContext.setRetryContext(status, dataSource);
+    return status;
+  }
+
+  /**
+   * Commits the transaction associated with the current {@link Thread} and clears the {@link RetryContext}.
+   * Can be called only within an established {@link RetryContext}.
+   * @throws TransactionException Forwarded from {@link PlatformTransactionManager#commit(TransactionStatus)}
+   * @throws IllegalArgumentException Thrown when called outside a {@link RetryContext}
+   */
+  public void commit() throws TransactionException {
+    try {
+      RetryContext context = RetryContext.getRetryContext();
+      transactionManagers.get(context.datasource).commit(context.transactionStatus);
+    } finally {
+      RetryContext.clearRetryContext();
+    }
+  }
+
+  /**
+   * Rollbacks the transaction associated with the current {@link Thread} and clears the {@link RetryContext}.
+   * Can be called only within an established {@link RetryContext}.
+   * @throws TransactionException Forwarded from {@link PlatformTransactionManager#rollback(TransactionStatus)}
+   * @throws IllegalArgumentException Thrown when called outside a {@link RetryContext}
+   */
+  public void rollback() throws TransactionException {
+    try {
+      RetryContext context = RetryContext.getRetryContext();
```

Comment:
context could be nested, how do we calculate the transactionStatus?

Reply:
In case of nested context, the inner context will be rolled back, while the outer remains untouched.